vault/token_store.go

```diff
@@ -1342,17 +1382,41 @@ func (ts *TokenStore) handleTidy(ctx context.Context, req *logical.Request, data
 			// Look up tainted entries so we can be sure that if this isn't
 			// found, it doesn't exist. Doing the following without locking
 			// since appropriate locks cannot be held with salted token IDs.
+			// Also perform deletion if the parent doesn't exist any more.
 			te, _ := ts.lookupSalted(ctx, child, true)
-			if te == nil {
+			// If the child entry is not nil, but the parent doesn't exist, then turn
+			// that child token into an orphan token. Theres no deletion in this case.
+			if te != nil && exists == nil {
+				lock := locksutil.LockForKey(ts.tokenLocks, te.ID)
+				lock.Lock()
+
+				te.Parent = ""
+				err = ts.store(ctx, te)
+				if err != nil {
+					tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to convert child token into an orphan token: %v", err))
+				}
+				lock.Unlock()
+				continue
+			}
+			// Otherwise, if the entry doesn't exist, or if the parent doesn't exit go
+			// on with the delete onthe secondary index
+			if te == nil || exists == nil {
 				index := parentPrefix + parent + child
 				ts.logger.Trace("token: deleting invalid secondary index", "index", index)
 				err = ts.view.Delete(ctx, index)
 				if err != nil {
 					tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to delete secondary index: %v", err))
+					continue
 				}
-				deletedCountParentList++
+				deletedChildrenCount++
 			}
 		}
+		// Add current children deleted count to the total count
+		deletedCountParentList += deletedChildrenCount
+		// If we deleted all the children, then add that to our deleted parent entries count
```

Comment:
Should the parent prefix be cleaned out here if `exists == nil`? Since ClearView above operates at revocation time but in this situation we won't be revoking the parent.

Reply:
Consul will remove the prefix entry if the "directory" is empty so I didn't have an explicit delete in here, but I guess we can't guarantee that this is the behavior for other storage backends so I might be good to explicitly call a delete (though I'd do it inside `if originalChildrenCount == deletedChildrenCount {...}` which ensures that we delete the prefix iff all child/secondary indexes are removed).

Reply:
👍